sdk/identity/azure-identity/azure/identity/_authn_client.py

```diff
@@ -213,30 +214,38 @@ def request_token(
         return token
 
     def obtain_token_by_refresh_token(self, scopes, username):
-        # type: (Iterable[str], str) -> Optional[AccessToken]
+        # type: (Iterable[str], Optional[str]) -> Optional[AccessToken]
         """Acquire an access token using a cached refresh token. Returns ``None`` when that fails, or the cache has no
         refresh token. This is only used by SharedTokenCacheCredential and isn't robust enough for anything else."""
 
-        # find account matching username
-        accounts = self._cache.find(TokenCache.CredentialType.ACCOUNT, query={"username": username})
-        for account in accounts:
-            # try each refresh token that might work, return the first access token acquired
-            for token in self.get_refresh_tokens(scopes, account):
-                # currently we only support login.microsoftonline.com, which has an alias login.windows.net
-                # TODO: this must change to support sovereign clouds
-                environment = account.get("environment")
-                if not environment or (environment not in self._auth_url and environment != "login.windows.net"):
-                    continue
-
-                request = self.get_refresh_token_grant_request(token, scopes)
-                request_time = int(time.time())
-                response = self._pipeline.run(request, stream=False)
-                try:
-                    return self._deserialize_and_cache_token(
-                        response=response, scopes=scopes, request_time=request_time
-                    )
-                except ClientAuthenticationError:
-                    continue
+        #if an username is provided, restrect our search to accounts that have that username
+        query = {} if username is None else {"username": username}
```

Comment:
Is it intentional that an empty string is different than `None`? If not, you can use the truthy:ness of the value rather than checking for `is None`" - e.g.

```python
query = {"username": username} if username else {}
```

Reply:
Thanks! To be honest - I'm not sure what the original intention of "" was, and just tried to not change any existing behavior.  I believe that previously always caused retrieval from the cache to fail, but treating it as another spelling of "if there is a single identity, please use it" does seem better.  I'll make the change based on my limited understanding of what is pythonic.